components/board/collector.go

```diff
@@ -26,78 +27,66 @@ func (m method) String() string {
 	return "Unknown"
 }
 
-// AnalogRecords a collection of AnalogRecord.
-type AnalogRecords struct {
-	Readings []AnalogRecord
-}
-
-// AnalogRecord a single analog reading.
-type AnalogRecord struct {
-	AnalogName  string
-	AnalogValue int
-}
-
-// GpioRecords a collection of GpioRecord.
-type GpioRecords struct {
-	Readings []GpioRecord
-}
-
-// GpioRecord a signle gpio reading.
-type GpioRecord struct {
-	GPIOName  string
-	GPIOValue bool
-}
-
-func newAnalogCollector(resource interface{}, params data.CollectorParams) (data.Collector, error) {
+// NewAnalogCollector returns a collector to register an analog reading method. If one is already registered
+// with the same MethodMetadata it will panic.
+func NewAnalogCollector(resource interface{}, params data.CollectorParams) (data.Collector, error) {
 	board, err := assertBoard(resource)
 	if err != nil {
 		return nil, err
 	}
 
 	cFunc := data.CaptureFunc(func(ctx context.Context, arg map[string]*anypb.Any) (interface{}, error) {
-		var readings []AnalogRecord
-		for k := range arg {
-			if reader, ok := board.AnalogReaderByName(k); ok {
-				value, err := reader.Read(ctx, data.FromDMExtraMap)
-				if err != nil {
-					// A modular filter component can be created to filter the readings from a component. The error ErrNoCaptureToStore
-					// is used in the datamanager to exclude readings from being captured and stored.
-					if errors.Is(err, data.ErrNoCaptureToStore) {
-						return nil, err
-					}
-					return nil, data.FailedToReadErr(params.ComponentName, analogs.String(), err)
+		var value int
+		if _, ok := arg["reader_name"]; !ok {
```

Comment:
Since this is repeated 3 times, extract into a const? Also, if you can add "Must supply `reader_name` in additional params for analog collector" to make it clear what the user needs to update

Reply:
It'd be good to also catch this at reconfiguration time. We can start validating [here](https://github.com/viamrobotics/rdk/blob/b34d7db3573c3bb9e0d7dbeddaa1c16407bf1b05/services/datamanager/builtin/builtin.go#L205-L216) by mapping `MethodMetadata` to expected params (locally on a fake board, you can print the method metadata to get what it should look like). Since this is the only collector that's using additional params right now, this would only contain `reader_name`.

Reply:
But even for the board collectors, one is `reader_name` the other is `pin_name`. So would I just need to enforce that additional params should contain one? Do I also need to enforce this on every collector or would it special case to only validate if it is a board component?

Reply:
MethodMetadata contains the component + method, so the board + reader would map to `reader_name` and board + gpio would map to `pin_name`. If the method metadata exists in the dict, you check that it contains the additional params listed. Let me know if that makes sense!